File: leetcode/Python/palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        dp, subsol, sol = [[0 for _ in range(len(s))] for _ in range(len(s))], list(), list()
        for d in range(len(s)):
            for i in range(len(s)-d):
                f = 1 if i+1 > i+d-1 else dp[i+1][i+d-1]
                dp[i][i+d] = 1 if f == 1 and s[i]==s[i+d] else 0
        def dfs(idx):
            if idx >= len(s):
                sol.append(subsol[:])
            for i in range(idx, len(s)):
                if dp[idx][i] == 1:
                    subsol.append(s[idx:i+1])
                    dfs(i+1)
                    subsol.pop()
        dfs(0)
        return sol

File: leetcode/C++/palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> sol;
    vector<string> subsol;
    vector<vector<int>> m;
    void dfs(int idx, string &s){
        if(idx>=s.length()) {
            sol.push_back(subsol);
            return;
        }
        for(int i=idx; i<s.length(); ++i){
            if(m[idx][i]) {
                subsol.push_back(s.substr(idx, i-idx+1));
                dfs(i+1, s);
                subsol.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        m = vector<vector<int>>(s.length(), vector<int>(s.length(),0));
        for(int d=0,f=0; d<s.length(); ++d){
            for(int i=0; i+d<s.length(); ++i){
                f = (i+1>i+d-1)? 1 : m[i+1][i+d-1];
                if(f && s[i]==s[i+d]) m[i][i+d]=1;
            }
        }
        dfs(0, s);
        return sol;
    }
};